nexus/db-queries/src/db/datastore/volume_repair.rs

```diff
@@ -39,7 +39,16 @@ impl DataStore {
             .execute_async(&*conn)
             .await
             .map(|_| ())
-            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+            .map_err(|e| {
+                if e.to_string().contains(
+                    "duplicate key value violates \
+                    unique constraint \"volume_repair_pkey\"",
+                ) {
```

Comment:
Could we, perhaps, check whether the error is `diesel::result::Error::DatabaseError` with [`DatabaseErrorKind::UniqueViolation`](https://docs.rs/diesel/latest/diesel/result/enum.DatabaseErrorKind.html#variant.UniqueViolation) and then check the value of the [`constraint_name` method on the `DatabaseErrorInformation`? 

That seems preferable to doing this based on the string representation: it avoids allocating the string (which is probably not a big deal here), but more importantly, it should be more resilient to a `diesel` change  that modified the formatted representation of the error. If diesel changed the `fmt::Display` implementation in a semver-compatible release, this code would silently break, but changing the API for programmatically inspecting the error type as I suggested would be a semver-breaking change.

Reply:
This is so much better, thank you! Done in 920724824